### PR TITLE
Grafana-UI: Space component

### PR DESCRIPTION
**Pull request by joshhunt**
**What this PR does / why we need it**:

A more flexible spacing primitive for when VerticalGroup/HorizontalGroup are not suitable.

Screenshot:

image

**Special notes for your reviewer**:

Later on it might be nice to have an element like `<Interpose node={<Space v={2} />{children}<Interpose>`, kind of like `Array.join()`, for an easier way to apply even spacing between elements :)

### CODE REVIEW

File: packages/grafana-ui/src/components/Space/Space.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import { css, cx } from '@emotion/css';
+import { GrafanaTheme2 } from '@grafana/data';
+import { useTheme2 } from '../../themes';
+
+export interface SpaceProps {
+  v?: number;
+  h?: number;
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/Space/Space.tsx:8]: Not _super_ enthusiastic about such a shorthand, but I think it's the easiest? Definitely open to vertical/horizontal or other suggestions :)

**Comment by ryantxu** [comment]
This is the most clear screenshot I have seen :rofl:  
image

**Comment by hugohaggmark** [comment]
We brought this PR up on our weekly PR and Bugs scrub meeting and we agreed that this is not a good idea.

File: packages/grafana-ui/src/components/Space/Space.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  return <span className={cx(styles.wrapper)} />;
+};
+
+Space.defaultProps = {
+  v: 0,
+  h: 0,
+  layout: 'block',
+};
+
+const getStyles = (theme: GrafanaTheme2, props: SpaceProps) => ({
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Space/Space.tsx:25]: use stylesFactory here to memoize the styles

File: packages/grafana-ui/src/components/Space/Space.tsx [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  const styles = getStyles(theme, props);
+
+  return <span className={cx(styles.wrapper)} />;
+};
+
+Space.defaultProps = {
+  v: 0,
+  h: 0,
+  layout: 'block',
+};
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Space/Space.tsx:23]: think using default function props is what we usually do for this 

```ts
export function Space({layout = 'block'}: SpaceProps) {
```

File: packages/grafana-ui/src/components/Space/Space.tsx (hunk 1 above)
Comment (torkelo) [reply packages/grafana-ui/src/components/Space/Space.tsx:8]: can we not use the same syntax as the spacing function in the theme? an array with 1-4 elements [top, right, bottom, left]

File: packages/grafana-ui/src/components/Space/Space.tsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import { GrafanaTheme2 } from '@grafana/data';
+import { useTheme2 } from '../../themes';
+
+export interface SpaceProps {
+  v?: number;
+  h?: number;
+  layout?: 'block' | 'inline';
+}
+
+export const Space = (props: SpaceProps) => {
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Space/Space.tsx:12]: not thrilled witih the name given the single namespace of grafana/ui 
alternatives
* LayoutSpace 
* LayoutSpacing

would probably then add some flex options as well, display: 'flex' | 'block' | 'inline' , etc and justifyContent & alignItems

File: packages/grafana-ui/src/components/Space/Space.tsx (hunk 1 above)
Comment (joshhunt) [reply packages/grafana-ui/src/components/Space/Space.tsx:8]: If you look at my example (vertical space between two block-level elements), using the spacing function syntax would make it `<Space space={[2,0,0,0]} />` which I think is a bit more cumbersome than it needs to be.

Did you have a use case in mind where the shorthand of `[2]` (4 units of spacing both horizontally vertically) or `[2, 0]` (2 units of spacing just vertically, but harder to read) would be useful?

File: packages/grafana-ui/src/components/Space/Space.tsx (hunk 4 above)
Comment (joshhunt) [reply packages/grafana-ui/src/components/Space/Space.tsx:12]: I think the name Space is fine :) It's simple, does the job, and don't think there's confusion about what it does. But also, it's just a name so I don't care if you feel more strongly in favor of something else.

I'm not sure if any flex options are appropriate though - his should not contain any children and should be used only to provide a fixed amount of space. There could be room for like a `<FlexItem grow={1} />` element that could contain children. 

I do think some other layout primitives would be useful, but I want to be careful in not overloading this one component, and making things that can be composed together easily.

**Pull request closed without merging**